File: server/lib/backup/exporter.py
import json
import os
import uuid
from zipfile import ZipFile

from sqlmodel import Session, select

import configure
from api.image import get_image_and_file_path_or_404
from database.sql_model import Room, Bot, Persona, Prompt, Conversation, Summary


class DataExporter:
    def __init__(self, session: Session, username: str):
        os.makedirs(configure.get_store_path(f'{username}/export'), exist_ok=True)
        self.uuid = uuid.uuid4().hex
        self.file = ZipFile(configure.get_store_path(f"{username}/export/{self.uuid}.zip"), "w")
        self.dirs = {}
        self.exported_rooms = {}
        self.exported_bots = {}
        self.exported_prompts = {}
        self.exported_personas = {}
        self.exported_images = {}
        self.exported_conversations = {}
        self.exported_summaries = {}

        self.session = session
        self.username = username

    def finish(self):
        self.file.writestr("meta.json", json.dumps({
            "rooms": list(self.exported_rooms.keys()),
            "bots": list(self.exported_bots.keys()),
            "personas": list(self.exported_personas.keys()),
            "prompts": list(self.exported_prompts.keys()),
            "images": list(self.exported_images.keys()),
            "conversations": list(self.exported_conversations.keys()),
            "summaries": list(self.exported_summaries.keys()),
        }))

        self.file.close()

    def mkdir_exists_ok(self, name):
        if name in self.dirs:
            return
        self.dirs[name] = True
        self.file.mkdir(name)

    def export_room(self, room: Room):
        if room.id in self.exported_rooms:
            return

        self.mkdir_exists_ok('rooms')
        self.file.writestr(f"rooms/{room.id}.json", room.model_dump_json())
        self.exported_rooms[room.id] = True

        if room.bot:
            self.export_bot(room.bot)

        if room.persona:
            self.export_persona(room.persona)

        if room.prompt:
            self.export_prompt(room.prompt)

        if room.summary_prompt:
            self.export_prompt(room.summary_prompt)

        if room.re_summary_prompt:
            self.export_prompt(room.re_summary_prompt)

        if room.translate_prompt:
            self.export_prompt(room.translate_prompt)

        for conversation in self.session.exec(select(Conversation).where(Conversation.room_id == room.id)):
            self.export_conversation(conversation)

        for summary in self.session.exec(select(Summary).where(Summary.room_id == room.id)):
            self.export_summary(summary)

    def export_bot(self, bot: Bot):
        if bot.id in self.exported_bots:
            return

        self.mkdir_exists_ok('bots')
        self.file.writestr(f"bots/{bot.id}.json", bot.model_dump_json())
        self.exported_bots[bot.id] = True
        if bot.profileImageId:
            self.export_image(bot.profileImageId)

        if bot.image_assets:
            image_assets = json.loads(bot.image_assets)
            for image_asset in image_assets:
                self.export_image(image_asset['imageId'])

    def export_persona(self, persona: Persona):
        if persona.id in self.exported_personas:
            return

        self.mkdir_exists_ok('personas')
        self.file.writestr(f"personas/{persona.id}.json", persona.model_dump_json())
        if persona.profileImageId:
            self.export_image(persona.profileImageId)

        self.exported_personas[persona.id] = True

    def export_prompt(self, prompt: Prompt):
        if prompt.id in self.exported_prompts:
            return

        self.mkdir_exists_ok('prompts')

        if prompt.llm_config:
            llm_config = json.loads(prompt.llm_config)
            if 'key' in llm_config:
                del llm_config['key']
            prompt.llm_config = json.dumps(llm_config)

        self.file.writestr(f"prompts/{prompt.id}.json", prompt.model_dump_json())
        self.exported_prompts[prompt.id] = True

    def export_image(self, image_id: str):
        if image_id in self.exported_images:
            return
        try:
            image, path = get_image_and_file_path_or_404(self.session, self.username, image_id)
            self.mkdir_exists_ok('images')
            self.file.writestr(f"images/{image_id}.json", image.model_dump_json())
            self.file.write(path, f"images/{image_id}.bin")
            self.exported_images[image_id] = True
        except:
            pass

    def export_conversation(self, conversation: Conversation):
        if conversation.id in self.exported_conversations:
            return

        self.mkdir_exists_ok('conversations')
        self.file.writestr(f"conversations/{conversation.id}.json", conversation.model_dump_json())
        self.exported_conversations[conversation.id] = True

    def export_summary(self, summary: Summary):
        if summary.id in self.exported_summaries:
            return

        self.mkdir_exists_ok('summaries')
        self.file.writestr(f"summaries/{summary.id}.json", summary.model_dump_json())
        self.exported_summaries[summary.id] = True

File: client/lib/data/Exporter.ts
"use client";

import * as fflate from 'fflate';
import {dumpRoom, RawRoom} from "@/lib/data/Room";
import Bot, {getBot} from "@/lib/data/bot/Bot";
import {buildImageLink, getImage} from "@/lib/data/Image";
import ImageAsset from "@/lib/data/bot/ImageAsset";
import Persona, {getPersona} from "@/lib/data/Persona";
import Prompt, {getPrompt} from "@/lib/data/Prompt";
import {RawConversation} from "@/lib/data/Conversation";
import {api} from "@/lib/api-client";
import Summary from "@/lib/data/Summary";
import {LoreBookReadResult} from "@/lib/data/lore/ReadLoreBook";

export class Exporter {
    zip: fflate.Zip;
    exportedRooms = new Set<string>();
    exportedBots = new Set<string>();
    exportedPrompts = new Set<string>();
    exportedPersonas = new Set<string>();
    exportedImages = new Set<string>();
    exportedConversations = new Set<string>();
    exportedSummaries = new Set<string>();
    exportedLorebooks = new Set<string>();

    private zipChunks: Uint8Array[] = [];
    setStatus: (status: string) => void = () => {
    };

    constructor(setStatus: (status: string) => void) {
        this.zip = new fflate.Zip((err, dat, final) => {
            if (err) {
            } else if (dat) {
                this.zipChunks.push(dat);
                if (final) {
                    const fullZipData = this.concatenateChunks(this.zipChunks);
                    this.downloadZip(fullZipData); // 최종 Zip 다운로드
                }
            }
        });
        this.setStatus = setStatus;
    }

    private concatenateChunks(chunks: Uint8Array[]): Uint8Array {
        // 모든 Uint8Array 데이터를 하나로 합치기
        const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
        const result = new Uint8Array(totalLength);
        let offset = 0;
        for (const chunk of chunks) {
            result.set(chunk, offset);
            offset += chunk.length;
        }
        return result;
    }

    finishAndDownload() {
        const metadata = {
            "version": 1,
            "rooms": [...this.exportedRooms],
            "bots": [...this.exportedBots],
            "prompts": [...this.exportedPrompts],
            "personas": [...this.exportedPersonas],
            "images": [...this.exportedImages],
            "conversations": [...this.exportedConversations],
            "summaries": [...this.exportedSummaries],
            "lorebooks": [...this.exportedLorebooks],
        }
        this.addFileToZip("meta.json", JSON.stringify(metadata));
        this.zip.end()
    }

    private downloadZip(data: Uint8Array) {
        const blob = new Blob([data], {type: "application/zip"});
        const url = URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = url;
        a.download = "exported.zip";
        a.click();
        URL.revokeObjectURL(url);
    }

    private addFileToZip(path: string, content: string | Uint8Array) {
        if (typeof content === "string") {
            const text = new fflate.ZipDeflate(path, {
                level: 9
            })
            this.zip.add(text);
            text.push(fflate.strToU8(content), true)
        } else {
            const binary = new fflate.ZipPassThrough(path)
            this.zip.add(binary);
            binary.push(content, true)
        }
    }

    private async fetchBinary(url: string): Promise<Uint8Array | null> {
        try {
            const res = await fetch(url);
            if (!res.ok) throw new Error(`Failed to fetch binary: ${url}`);
            return new Uint8Array(await res.arrayBuffer());
        } catch (e) {
            console.error(e);
            return null;
        }
    }

    async exportRoom(room_id: string, include_conversation: boolean) {
        if (this.exportedRooms.has(room_id)) return;

        const room: RawRoom = await dumpRoom(room_id)

        await this.exportRawRoom(room, include_conversation)
    }

    async exportRawRoom(room: RawRoom, include_conversation: boolean) {
        if (this.exportedRooms.has(room.id)) return;

        this.setStatus(`방 내보내기: ${room.name}`)
        this.addFileToZip(`rooms/${room.id}.json`, JSON.stringify(room));
        if (room.bot_id) await this.exportBot(room.bot_id);

        if (room.persona_id) await this.exportPersona(room.persona_id);

        if (room.prompt_id) await this.exportPrompt(room.prompt_id);

        if (room.summary_prompt_id) await this.exportPrompt(room.summary_prompt_id);

        if (room.re_summary_prompt_id) await this.exportPrompt(room.re_summary_prompt_id);

        if (room.translate_prompt_id) await this.exportPrompt(room.translate_prompt_id);

        if (include_conversation) {
            const conversations: RawConversation[] = await api('room/' + room.id + '/conversation/dump', {
                method: 'GET'
            })
            for (const conversation of conversations) {
                await this.exportConversation(conversation)
            }

            const summaries: Summary[] = await api('room/' + room.id + '/summary/dump', {
                method: 'GET'
            })

            for (const summary of summaries) {
                await this.exportSummary(summary)
            }
        }

        this.exportedRooms.add(room.id);
    }

    async exportBot(bot_id: string) {
        if (this.exportedBots.has(bot_id)) return;
        const bot = await getBot(bot_id)

        if (bot.lore_book_id) {
            await this.exportLoreBook(bot.lore_book_id)
        }

        await this.exportRawBot(bot)
    }

    async exportLoreBook(lore_id: string) {
        if (this.exportedLorebooks.has(lore_id)) return;
        const lore: LoreBookReadResult = await api(`lorebook/${lore_id}/read`, {
            method: 'GET'
        })

        this.setStatus(`로어북 내보내기: ${lore.book.name}`)
        this.addFileToZip(`lorebooks/${lore_id}.json`, JSON.stringify(lore));
        this.exportedLorebooks.add(lore_id);
    }

    async exportRawBot(bot: Bot) {
        this.setStatus(`봇 내보내기: ${bot.name}`)
        this.addFileToZip(`bots/${bot.id}.json`, JSON.stringify(bot));

        if (bot.profileImageId) {
            await this.exportImage(bot.profileImageId)
        }

        if (bot.image_assets) {
            const assets: ImageAsset[] = JSON.parse(bot.image_assets);
            for (const asset of assets) {
                await this.exportImage(asset.imageId)
            }
        }

        this.exportedBots.add(bot.id);
    }

    async exportPersona(persona_id: string) {
        if (this.exportedPersonas.has(persona_id)) return;

        const persona = await getPersona(persona_id)

        await this.exportRawPersona(persona)
    }

    async exportRawPersona(persona: Persona) {
        this.setStatus(`페르소나 내보내기: ${persona.name}`)

        this.addFileToZip(`personas/${persona.id}.json`, JSON.stringify(persona));
        if (persona.profileImageId) {
            await this.exportImage(persona.profileImageId)
        }
        this.exportedPersonas.add(persona.id);
    }

    async exportPrompt(prompt_id: string) {
        if (this.exportedPrompts.has(prompt_id)) return;

        const prompt = await getPrompt(prompt_id)

        await this.exportRawPrompt(prompt)
    }

    async exportRawPrompt(prompt: Prompt) {
        this.setStatus(`프롬프트 내보내기: ${prompt.name}`)

        const cloned: Prompt = {...prompt};

        if (cloned.llm_config) {
            const llmConfig = JSON.parse(cloned.llm_config);
            if (llmConfig.key) {
                llmConfig.key = "";
                cloned.llm_config = JSON.stringify(llmConfig);
            }
        }

        this.addFileToZip(`prompts/${prompt.id}.json`, JSON.stringify(cloned));
        this.exportedPrompts.add(prompt.id);
    }

    async exportImage(id: string) {
        if (this.exportedImages.has(id)) return;

        this.setStatus(`이미지 내보내기: ${id}`)

        const image = await getImage(id)
        if (!image) return;
        const binary = await this.fetchBinary(buildImageLink(id, 'original')!)
        if (!binary) return;
        this.addFileToZip(`images/${id}.bin`, binary);
        this.addFileToZip(`images/${id}.json`, JSON.stringify(image));
        this.exportedImages.add(id);
    }

    async exportConversation(conversation: RawConversation) {
        if (this.exportedConversations.has(conversation.id)) return;
        this.setStatus(`대화 내보내기: ${conversation.id}`)
        this.addFileToZip(`conversations/${conversation.id}.json`, JSON.stringify(conversation));
        this.exportedConversations.add(conversation.id);
    }

    async exportSummary(summary: Summary) {
        if (this.exportedSummaries.has(summary.id)) return;
        this.setStatus(`요약 내보내기: ${summary.id}`)
        this.addFileToZip(`summaries/${summary.id}.json`, JSON.stringify(summary));
        this.exportedSummaries.add(summary.id);
    }
}